### PR TITLE
Clamp scale with lower and upper bounds

### PR DESCRIPTION
**Pull request by OliBomby**
This is mostly a code quality upgrade.
Reworked `ClampScaleToPlayfieldBounds` to clamp the scale value to both a upper bound and an upper bound. Previously only did an upper bound. With this it supports negative scale values. I also added comments to make it more clear what is going on there.

There is this rare situation where the lower bound equals 1, previously it would allow you to scale this by 0.5 which would direct the circle off-screen.
osu Game Rulesets Osu Tests_Y9keZwEQY8

### CODE REVIEW

**Comment by bdach** [comment]
I'm kinda confused by this PR in general.

> With this it supports negative scale values.

Is this used anywhere?

> This is mostly a code quality upgrade.

It is? Even in light of the previous quoted statement?

> There is this rare situation where the lower bound equals 1, previously it would allow you to scale this by 0.5 which would direct the circle off-screen.

In what circumstances is this possible?

File: osu.Game.Rulesets.Osu/Edit/OsuSelectionScaleHandler.cs [hunk 1]
```diff
@@ -240,39 +240,66 @@ public Vector2 ClampScaleToPlayfieldBounds(Vector2 scale, Vector2? origin = null
...
-            }
+                scale = clampToBound(scale, point, Vector2.Zero, OsuPlayfield.BASE_SIZE);
 
-            return Vector2.ComponentMax(scale, new Vector2(Precision.FLOAT_EPSILON));
-
-            float minPositiveComponent(Vector2 v) => MathF.Min(v.X < 0 ? float.PositiveInfinity : v.X, v.Y < 0 ? float.PositiveInfinity : v.Y);
+            return scale;
 
-            Vector2 clampToBound(Vector2 s, Vector2 p, Vector2 bound)
+            Vector2 clampToBound(Vector2 s, Vector2 p, Vector2 lowerBounds, Vector2 upperBounds)
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Edit/OsuSelectionScaleHandler.cs:247]: If we're working on code quality here - what is `s`? what is `p`? rename, or xmldoc?

File: osu.Game.Rulesets.Osu/Edit/OsuSelectionScaleHandler.cs [hunk 2]
```diff
@@ -240,39 +240,66 @@ public Vector2 ClampScaleToPlayfieldBounds(Vector2 scale, Vector2? origin = null
...
+                var sLowerBounds = Vector2.Divide(lowerBounds, p);
+                var sUpperBounds = Vector2.Divide(upperBounds, p);
+                if (p.X < 0)
+                    (sLowerBounds.X, sUpperBounds.X) = (sUpperBounds.X, sLowerBounds.X);
+                if (p.Y < 0)
+                    (sLowerBounds.Y, sUpperBounds.Y) = (sUpperBounds.Y, sLowerBounds.Y);
+                if (Precision.AlmostEquals(p.X, 0))
+                    (sLowerBounds.X, sUpperBounds.X) = (float.NegativeInfinity, float.PositiveInfinity);
+                if (Precision.AlmostEquals(p.Y, 0))
+                    (sLowerBounds.Y, sUpperBounds.Y) = (float.NegativeInfinity, float.PositiveInfinity);
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Edit/OsuSelectionScaleHandler.cs:300]: What is this?

**Comment by OliBomby** [comment]
> > With this it supports negative scale values.
> 
> Is this used anywhere?

Currently not yet, but Marvin considered adding negative scaling to the selection box which is WIP 
 https://github.com/minetoblend/osu/tree/feature/select-box-negative-scale

> > This is mostly a code quality upgrade.
> 
> It is? Even in light of the previous quoted statement?

Maybe not entirely accurate statement. It started as a slight feature enhancement for the lower bound stuff, but in the end the code quality became a lot more self-explaining, so I think its a code quality upgrade too.

> > There is this rare situation where the lower bound equals 1, previously it would allow you to scale this by 0.5 which would direct the circle off-screen.
> 
> In what circumstances is this possible?

It's the circumstances shown in the screenshot. It has a rotated grid and scaling on one axis, so a decrease in scaling moves the circle straight towards the grid's X axis which increases the Y coordinate out of bounds.

**Review by bdach — APPROVED** [review]

**Pull request merged**